src/systems/environmental_sensor_system/EnvironmentalSensorSystem.cc

```diff
@@ -0,0 +1,349 @@
+  /// \brief Update the sensor and generate data
+  /// \param[in] _now The current time
+  /// \return True if the update was successful
+  public: virtual bool Update(
+    const std::chrono::steady_clock::duration &_now) override
+  {
+    if (!this->ready) return false;
+
+    if (!this->session.has_value()) return false;
+
+    this->session = this->gridField->frame[this->field].StepTo(
+      this->session.value(), std::chrono::duration<double>(_now).count());
+
+    if (!this->session.has_value()) return false;
+
+    gz::msgs::Double msg;
+    *msg.mutable_header()->mutable_stamp() = gz::msgs::Convert(_now);
+    auto frame = msg.mutable_header()->add_data();
+    frame->set_key("frame_id");
+    frame->add_value(this->Name());
+    auto data = this->gridField->frame[this->field].LookUp(
+      this->session.value(), this->position);
+    if (!data.has_value())
+      return false;
+    msg.set_data(data.value());
+    // TODO(anyone) Add sensor noise.
+    this->pub.Publish(msg);
+    return true;
+  }
+
+  ////////////////////////////////////////////////////////////////
+  /// \brief Attempts to set a data table.
+  /// \param[in] _data - The data table
+  /// \param[in] _curr_time - The current time.
+  public: void SetDataTable(
+    const components::Environment* _data,
+    const std::chrono::steady_clock::duration &_curr_time)
+  {
+    gzdbg << "Setting new data table\n";
+    auto data = _data->Data();
+    if(!data->frame.Has(this->field))
+    {
+      gzwarn << "Environmental sensor could not find field "
+        << this->field << "\n";
+      ready = false;
+      return;
+    }
+
+    this->gridField = data;
+    this->session = this->gridField->frame[this->field].CreateSession();
+    this->session = this->gridField->frame[this->field].StepTo(
+      *this->session,
+      std::chrono::duration<double>(_curr_time).count());
+    this->ready = true;
+
+    if(!this->session.has_value())
+    {
+      gzerr << "Exceeded time stamp." << std::endl;
+    }
+  }
+
+  ////////////////////////////////////////////////////////////////
+  /// \brief Updates the position of the sensor at each time step
+  /// to track a given entity,
+  /// \param[in] - The sensor entity.
+  /// \param[in] - The ECM to use for tracking.
+  public: bool UpdatePosition(
+    const Entity _entity,
+    const EntityComponentManager& _ecm)
+  {
+    if (this->gridField->reference != math::SphericalCoordinates::SPHERICAL)
+    {
+      auto position = worldPose(_entity, _ecm).Pos();
+      this->position = position;
+      return true;
+    }
+    else
+    {
+      auto coordinates = gz::sim::sphericalCoordinates(_entity, _ecm);
+      if (!coordinates.has_value())
+      {
+        gzerr << "Sensor data is in spherical coordinates,"
+          << " but world has no spherical coordinates set.\n";
+        return false;
+      }
+
+      this->position = coordinates.value();
+      return true;
+    }
```

Comment:
@arjo129 consider:

```c++
const auto position = worldPose(_entity, _ecm);
if (this->gridField->reference != math::SphericalCoordinates::LOCAL2)
{
    auto origin = _ecm.Component<components::SphericalCoordinates>(worldEntity(_ecm));
    if (!origin) 
    {
        return false;
    }
    this->position = origin->Data().PositionTransform(
        position, math::SphericalCoordinates::LOCAL2, this->gridField->reference);
}
else
{
    this->position = position;
}
return true;
```

This will always work no matter what is the data reference system. I should do the same for the DVL :sweat_smile: (over there I assume an spherical coordinates component will always be around, which isn't true).